src/goto-symex/symex_complexity.h

```diff
@@ -0,0 +1,155 @@
+/// bounds due to internal structures becoming too large or solving conditions
+/// becoming too big for the SAT solver to deal with.
+///
+/// On top of the branch cancelling there is special consideration for loops.
+/// As branches get cancelled it keeps track of what loops are currently active
+/// up through the stack. If a loop has too many of its child branches killed
+/// (including additional loops or recursion) it won't do any more unwinds of
+/// that particular loop and blacklist it.
+///
+/// This blacklist is only held for as long as any loop is still active. As soon
+/// as the loop iteration ends and the context in which the code is being
+/// executed changes it'll be able to be run again.
+///
+/// Example of loop blacklisting:
+///
+///     loop A: (complexity: 5070)
+///         loop B: (complexity: 5000)
+///         loop C: (complexity: 70)
+///
+/// In the above loop B will be blacklisted if we have a complexity limitation
+/// < 5000, but loop A and B will still be run, because when loop B is removed
+/// the complexity of the loop as a whole is acceptable.
+class symex_complexityt
+{
+public:
+  symex_complexityt() = default;
+
+  explicit symex_complexityt(message_handlert &logger) : log(logger)
+  {
+  }
+
+  symex_complexityt(message_handlert &logger, const optionst &options)
+    : log(logger)
+  {
+    std::size_t limit = options.get_signed_int_option("symex-complexity-limit");
+    if((complexity_active = limit > 0))
+    {
+      // This gives a curve that allows low limits to be rightly restrictive,
+      // while larger numbers are very large to represent what you're asking.
+      max_complexity = static_cast<std::size_t>((limit * limit) * 75);
+
+      std::size_t failed_children_limit =
+        options.get_signed_int_option("symex-complexity-failed-children-limit");
+      std::size_t unwind = options.get_signed_int_option("unwind");
+
+      // In order of priority:
+      //  * explicit limit
+      //  * inferred limit from unwind
+      //  * best limit we can apply with very little information.
+      if(failed_children_limit > 0)
+        max_loops_complexity = failed_children_limit;
+      else if(unwind > 0)
+        max_loops_complexity = std::max(
+          static_cast<std::size_t>(std::floor(unwind / 3)),
+          static_cast<std::size_t>(1));
+      else
+        max_loops_complexity = limit;
+    }
+  }
+
+protected:
+  mutable messaget log;
+
+  /// Is the complexity module active, usually coincides with a max_complexity
+  /// value above 0.
+  bool complexity_active = false;
+
+  /// Functions called when the heuristic has been violated. Can perform
+  /// any form of branch, state or full-program transformation.
+  std::vector<symex_heuristic_transformationt> violation_transformations;
+
+  /// Default heuristic transformation. Adds unconditional false guards to
+  /// state.
+  symex_heuristic_transformationt default_transformation;
```

Comment:
Is there something else than the default? What is the procedure for choosing a different heuristic? Should this class be a template argument? Why do we have an instance of a class which has no field?

Reply:
1. No.
2. That's up for future people to decide. The entirely abstract design is this: you use [specified] heuristic that determines when a certain violation happens, which then runs a selection of [transformations] that can do many things: modify guards, modify the equation, modify all sorts of data structures, but all with the goal to ease the burden of whatever's happening. It's doesn't live up to that design right now, because I wanted to get things working before I went all-in, then I could decide from there how far I wanted to take it.
3. Don't think that's needed.
4. Is your point here that it's effectively a static method that doesn't really need an instance of anything, so why going through the trouble of creating an object at all? While that's correct, I prefer the idea that even the default has to adhere to the same API as all other transformations. This is a personal preference thing.

Reply:
I think this is alright, but it looks like you are trying to make the design be adaptable for a very generic class of heuristics, while there is no need for such customization level in the current implementation. This makes reasoning about the class currently more difficult than it would need to be, without being sure that it will not need to be changed in the future. 
I would prefer the design to start simple and close to what the first implementation needs (for instance make it a function instead of an abstract class) and then generalize as needed, (but you don't need to change it now).

Reply:
I prefer it being simple but also having the frameworks to support it in the future. It's the balance between building for right now and giving people hints and hooks to extend it in the future, while not wasting time building things that people will never use.

I don't mind simplifying it, as I was intending too anyway, but I'll still be keeping to that ideal either way.